### ISSUE TITLE
Support custom headers for CORS

### ISSUE THREAD

**Issue by alexandrevicenzi** [tus/tusd#135]
I'm using the oficial JS client with the oficial go server, the problem is that I need to send a custom header when it's JS code, and the [tusd middleware](https://github.com/tus/tusd/blob/master/unrouted_handler.go#L140) has fixed CORS headers.

It would be good to have a way to set custom CORS headers.

**Comment by Acconut** [comment]
I would recommend you to use a proxy in front of tusd which sets these headers, such as nginx or Apache. This allows for more complex header constructions, than tusd could ever provide.

On the other hand, if you use tusd as a Go package, this discussion shows you how to modify the CORS headers: https://github.com/tus/tusd/pull/123

**Comment by alexandrevicenzi** [comment]
I'm using nginx and I also use tuds as a go package.

The problem is that you can't set `Access-Control-Allow-Origin` with ` *` for example, because tuds middleware adds `Access-Control-Allow-Origin` with the current `Origin` header.

I don't fell that assuming fixed CORS headers are a good solution, it could have it as default, but it would be good to allow changing it.

**Comment by Acconut** [comment]
> The problem is that you can't set Access-Control-Allow-Origin with * for example, because tuds middleware adds Access-Control-Allow-Origin with the current Origin header.

Why would you want to use `*` as the value? Using the Origin header should be sufficient as it indicates the domain of the requesting site.

> I don't fell that assuming fixed CORS headers is a good solution, it could have it as default, but it would be good to allow changing it.

I disagree. IMO, it's not tusd's responsibility to provide the headers for the rest of your application. Are you having issues with setting the headers using nginx?

**Comment by alexandrevicenzi** [comment]
`*` was just an example.

What I want is to disable CORS on tusd because I want to set all CORS headers on nginx. I can't set `Access-Control-Allow-Origin` twice for example.

For example, my server doesn't allow DELETE requests, and the default CORS config allows it. It's ok to set all CORS headers if you use tuds as a client, but as a library I see this as a limitation.

**Comment by Acconut** [comment]
> What I want is to disable CORS on tusd because I want to set all CORS headers on nginx. I can't set Access-Control-Allow-Origin twice for example.

Nginx should allow you to overwrite headers which are already set, see https://stackoverflow.com/questions/14501047/how-to-add-a-response-header-on-nginx-when-using-proxy-pass

> For example, my server doesn't allow DELETE requests, and the default CORS config allows it.

I agree that this is suboptimal but as far as I know this should not cause any technical issues or does it?

**Comment by alexandrevicenzi** [comment]
To overwrite headers with nginx I need an extra module that isn't part of nginx. This won't work for me.

It doesn't cause any problem at all, it's just something that I would like to take control of.

My tusd server has OAuth, so it's a little bit hard to customize things sometimes. It would be nice to have some extra features to allow to use tusd as a customizable library.

I need to wrap things and do some hacks to get things working because I have a middleware before "tusd hadlers" to check if the user is authenticated.

For example, here is a small code of what I do:

```go
mux.Handle("/uploads/", func(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if origin := r.Header.Get("Origin"); origin != "" {
			header := w.Header()
			header.Set("Access-Control-Allow-Origin", origin)

			if r.Method == "OPTIONS" {
				// Preflight request
				header.Add("Access-Control-Allow-Methods", "POST, GET, HEAD, PATCH, DELETE, OPTIONS")
				header.Add("Access-Control-Allow-Headers", "Origin, Authorization, X-Requested-With, Content-Type, Upload-Length, Upload-Offset, Tus-Resumable, Upload-Metadata")
				header.Set("Access-Control-Max-Age", "86400")
			} else {
				// Actual request
				header.Add("Access-Control-Expose-Headers", "Upload-Offset, Location, Upload-Length, Tus-Version, Tus-Resumable, Tus-Max-Size, Tus-Extension, Upload-Metadata")
			}

			// Delete Origin to disable tusd CORS,
			// because we already set all CORS headers needed.
			r.Header.Del("Origin")
		}

		if r.Method == "OPTIONS" {
			next.ServeHTTP(w, r)
		} else {
			if ok := UserAuthenticated(r); ok {
				next.ServeHTTP(w, r)
			} else {
				w.WriteHeader(http.StatusUnauthorized)
			}
		}
	})
}(http.StripPrefix("/uploads/", tusHandler)))
```

**Comment by andrewmclagan** [comment]
I agree with the OP of this issue, its difficult to get this running with CORS as it currently stands. I dont understand why tusd needs to be adding CORS headers when its in proxy mode?

**Comment by Acconut** [comment]
> I dont understand why tusd needs to be adding CORS headers when its in proxy mode?

We use tusd behind a proxy and rely on tusd setting the correct CORS headers as it knows best which headers a necessary. So I think the situation depends highly on your setup.

So, @andrewmclagan, what are you suggesting? Would a flag to not set any CORS headers help? If so, we are happy to accept a PR for this.

**Comment by andrewmclagan** [comment]
Hello :)

I may be wrong in assuming this, although generally its been my experience that when using a forward proxy there is an assumption that all network related headers would be over-taken by the proxy? Whats your thoughts / approach?

**Comment by Acconut** [comment]
> I may be wrong in assuming this, although generally its been my experience that when using a forward proxy there is an assumption that all network related headers would be over-taken by the proxy?

Personally, I have never heard of such a rule but that doesn't mean it's not a best practice. What is your "definition" of network related headers? For me, this refers to headers such as `Connection: close` or `Cache-Control`. Both of which should (in most situation) be handled by the proxy.

But the CORS headers (`Access-Control-*`) are related to the application rather than the network in my mind. For example, the application knows best what headers it can safely expose to the browser (i.e. Access-Control-Expose-Headers) and this is not the proxies' matter. Of course, the proxy can still modify and extend these headers if it needs to do so for supporting other parts of the system but this is not tusd's job either.

For me, both tusd and the proxy can be responsible for setting the CORS headers and this responsibility is shared between these two components.

**Comment by andrewmclagan** [comment]
That seems like acceptable login to me

**Comment by Acconut** [comment]
This is implemented in #997 and will be included in the next release.
